### PR TITLE
plugin/kubernetes: do endpoint/slice check in retry loop

### PR DESCRIPTION
**Pull request by chrisohaver**
Signed-off-by: Chris O'Haver <cohaver@infoblox.com>

### 1. Why is this pull request needed and what does it do?

 If the API server is not available when CoreDNS starts, it may [not select the correct endpoint object type to watch](https://github.com/coredns/coredns/pull/4490#issuecomment-786072526).  When the wrong endpoint type is watched, DNS records for endpoints and headless services will be incomplete/missing.

This PR does the following:
- Moves the endpointslice check logic into a retry loop.
- Puts the code into a separate function so it's easier to identify and remove when the time comes.
- Documents reasoning and assumptions for the version check.

### 2. Which issues (if any) are related?

#4490 #4338

### CODE REVIEW

File: plugin/kubernetes/setup.go
Comment (miekg) [inline plugin/kubernetes/setup.go]: this is now a data race on k.opts.useEndpointSlices

File: plugin/kubernetes/setup.go
Comment (miekg) [inline plugin/kubernetes/setup.go]: ticker should be stopped when leaving the function

File: plugin/kubernetes/setup.go
Comment (miekg) [inline plugin/kubernetes/setup.go]: better to return a startup function that to keep passing caddy.Controllers

**Comment by chrisohaver** [comment]
I reworked the code with the frame of mind that watching EndpointSlices is normal behavior, and watching Endpoints is a special exception. `newdnsController` now will always define an EndpointSlice lister and controller.
Later, during coredns start up, but before starting the api controllers, we check with the api server to see if we need to watch Endpoints instead, and replace controllers if necessary.  In the future this step can be removed, and `newdnsController` can remain as is.

**Comment by chrisohaver** [comment]
If there is reluctance to implement endpointslice API support checks in a retry loop  ...

An alternative is to expose endpointslice use as a configurable option, and leave it to the user to configure which to use (defaulting to endpointslice).  This is more adaptable to custom k8s deployments, since a cluster can disable parts of endpointslice support via featuregate, which we don't currently detect (and I think difficult to detect dynamically).  It removes all k8s version checking from the code, but would require users of k8s < 1.19 to explicitly set the option. It adds yet-another-option to kubernetes, but the option could be deprecated later when we no longer support api.Endpoints.

Thoughts?

**Comment by miekg** [comment]
I'm in favor of checking for the users and not pushing this down to them. We also introducing an option which then needs to be removed again with all the config pain associated with that.

File: plugin/kubernetes/kubernetes.go
Comment (miekg) [inline plugin/kubernetes/kubernetes.go:301]: if we know what to do we can stop this loop, no?

Comment (miekg) [reply plugin/kubernetes/kubernetes.go:301]: oh we do, vnm

File: plugin/kubernetes/kubernetes.go
Comment (miekg) [inline plugin/kubernetes/kubernetes.go:278]: log.Debug/log.Warning ?

**Comment by miekg** [comment]
and some documentation updates (although it's getting a long README), but still wouldn't hurt

**Comment by chrisohaver** [comment]
Looking at the code with fresh eyes - I see that this introduces a race condition on `epController`. 

edit: ... and `epLister`

File: plugin/kubernetes/kubernetes.go
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go:278]: Trying to be concise with `starting server with unsynced Kubernetes API` ...

It's hard to be concise about the repercussions... which are that we continue to attempt to connect and sync in the background.  And until the API completely syncs for all watched objects, we will answer with SERVFAIL for any kubernetes query that does not resolve to a known record.

**Review by miekg — APPROVED** [review]
I think this looks OK

**Comment by chrisohaver** [comment]
Added sections to read me to describe the (pre-existing) 5 second start up delay behavior, and the endpointslice/endpoint monitor behavior.

I'd like to test this more before merging especially after the locking changes in ab47772, though I have not had time to focus on this in the past week.

**Pull request merged**
